README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
miRTarget : Integrative analysis of miRNA and mRNA
==================================================

miRTarget helps to find miRNA predicted and validated targets. Two databases are used [miRecords](http://c1.accurascience.com/miRecords/) and [miRTarBase](http://mirtarbase.mbc.nctu.edu.tw)

Installation and loading
------------------------

-   Install the latest version from [GitHub](https://github.com/kassambara/miRTarget) as follow:

``` r
# Install
if(!require(devtools)) install.packages("devtools")
devtools::install_github("kassambara/miRTarget")
```

-   Loading

``` r
library(miRTarget)
```

Examples of miRNAs
------------------

``` r
miRNAs <- c("hsa-miR-28-5p", "hsa-miR-768-5p")
```

Find miRNA targets
------------------

### Get predicted miRNA targets

[miRecords.v4](c1.accurascience.com/miRecords/) database is used. For each miRNA, we want to identify the predicted target genes by at least 5/11 databases.

The 11 databases are : "diana", "microinspector", "miranda", "mirtarget2", "mitarget", "nbmirtar", "pictar", "pita", "rna22", "rnahybrid", "targetscan".

Internet connexion is required.

``` r
p_targets <- get_mirecords_pt(miRNAs, n = 5, species = "Homo sapiens",
                                      result_file = "mirna_mirecords_predicted_targets.txt",
                              show_progress = FALSE)
#> Results have been saved in the following file: mirna_mirecords_predicted_targets.txt
head(p_targets, 10)
#>       mirna_name target.gene_name    status
#> 1  hsa-miR-28-5p              EN2 predicted
#> 2  hsa-miR-28-5p         KIAA0355 predicted
#> 3  hsa-miR-28-5p            MYST3 predicted
#> 4  hsa-miR-28-5p            HOXA1 predicted
#> 5  hsa-miR-28-5p            YPEL3 predicted
#> 6  hsa-miR-28-5p              RHO predicted
#> 7  hsa-miR-28-5p            LSM12 predicted
#> 9  hsa-miR-28-5p            IKZF4 predicted
#> 10 hsa-miR-28-5p          CSNK1G1 predicted
#> 11 hsa-miR-28-5p            CABP7 predicted
```

### Get miRNA validated targets

miRecords and miRtarbase are used. Internet connexion is required.

1.  miRecords

``` r
mirecords_vt <- get_mirecords_vt(miRNAs, 
                                 species = "Homo sapiens", version = 4)
#> Downloading miRecords miRNA validated targets...
head(mirecords_vt, 10)
#>       mirna_name target.gene_name    status
#> 1  hsa-miR-28-5p          gag-pol validated
#> 2 hsa-miR-768-5p             <NA> validated
```

1.  miRtarbase

``` r
mirtarbase_vt <- get_mirtarbase_vt(miRNAs)
#> Downloading miRTarbase miRNA validated targets...
head(mirtarbase_vt, 10)
#>       mirna_name target.gene_name    status
#> 1  hsa-miR-28-5p        LINC00346 validated
#> 2  hsa-miR-28-5p           SEMA4C validated
#> 3  hsa-miR-28-5p           ZNF106 validated
#> 4  hsa-miR-28-5p          NKIRAS2 validated
#> 5  hsa-miR-28-5p           IMPDH1 validated
#> 6  hsa-miR-28-5p             AAAS validated
#> 7  hsa-miR-28-5p           CDKN1A validated
#> 8  hsa-miR-28-5p            CENPV validated
#> 9  hsa-miR-28-5p            SMYD1 validated
#> 10 hsa-miR-28-5p            PAQR5 validated
```

1.  Combine validated targets

``` r
v_targets <- combine_mir_vt(mirecords_vt, mirtarbase_vt)
# head(v_targets) 
```

### Merge predicted and validated targets

``` r
targets <- combine_mir_pvt(p_targets, v_targets)
head(targets)
#>      mirna_name target.gene_name status
#> 1 hsa-miR-28-5p          gag-pol      V
#> 2 hsa-miR-28-5p        LINC00346      V
#> 3 hsa-miR-28-5p           SEMA4C      V
#> 4 hsa-miR-28-5p           ZNF106      V
#> 5 hsa-miR-28-5p          NKIRAS2      V
#> 6 hsa-miR-28-5p           IMPDH1      V
```

Correlation between miRNA expression and target gene expressions
----------------------------------------------------------------

### Expression values of miRNAs

``` r
# Create some data with 4 samples (S) and two miRNAs
mirna_exprs <- data.frame(S1 = c(99, 8690), S2 = c(31, 612),
                          S3 = c(33, 1747), S4 = c(29, 2540))
rownames(mirna_exprs) <- c("hsa-miR-28-5p", "hsa-miR-768-5p")

print(mirna_exprs)
#>                  S1  S2   S3   S4
#> hsa-miR-28-5p    99  31   33   29
#> hsa-miR-768-5p 8690 612 1747 2540
```

### Expression value of miRNA potential target genes

``` r
data("gene_exprs")
head(gene_exprs, 10)
#>            S1   S2   S3   S4
#> AFF3     2254   14   15   15
#> IL7       250  104   22   14
#> SIGLEC10 1258  221  106  110
#> TUBB2A   1678  484   71   48
#> UNG       130 1883  212  125
#> TPM4      424 3835  744  199
#> E2F6       66  389  456  622
#> HSPA6    1021   91   75  109
#> UMPS      165 1792  754  688
#> E2F3      361 2455 1388 1361
```

### Keep only targets, which gene expressions are available

``` r
# Keep only target with gene expression data
targets <- subset(targets, targets$target.gene_name %in% rownames(gene_exprs) )
# Order data by miRNA name
targets <- targets[order(targets$mirna_name),  ]
# Change rownames
rownames(targets) <- paste0(targets$mirna_name, "::", targets$target.gene_name)
head(targets)
#>                            mirna_name target.gene_name status
#> hsa-miR-28-5p::CENPV    hsa-miR-28-5p            CENPV      V
#> hsa-miR-28-5p::PRDM2    hsa-miR-28-5p            PRDM2      V
#> hsa-miR-28-5p::ARHGAP42 hsa-miR-28-5p         ARHGAP42      V
#> hsa-miR-28-5p::PRSS16   hsa-miR-28-5p           PRSS16      V
#> hsa-miR-28-5p::TUBB2A   hsa-miR-28-5p           TUBB2A      V
#> hsa-miR-28-5p::PCYOX1   hsa-miR-28-5p           PCYOX1      V
```

### Combine miRNA and mRNA expression data

The name and the number of samples in the two data set should be identical.

``` r
mir_mrna_exprs <- combine_mir_mrna_exprs(mirna_exprs, gene_exprs)
```

### Correlation between miRNAs and the corresponding target genes

> If a gene is targeted by a miRNA we expect that the expression profile of the gene and the miRNA are anti-correlated.

``` r
res.cor <- corr_mir_mrna( mir_targets = MirTarget(targets),
                          mir_mrna_exprs = mir_mrna_exprs, 
                          show_progress = FALSE, stand = TRUE)
```

-   Number of mRNAs per miRNAs with an absolute correlation of 0.6

``` r
viz_mrna_per_mir(res.cor, coeff = 0.6)
```

![](README-unnamed-chunk-15-1.png)

-   If you want to visualize the fraction of negative correlation use this

``` r
viz_neg_cor_fraction(res.cor)
```

![](README-unnamed-chunk-16-1.png)

### Keep only negative correlation

r \< -0.6

``` r
# Negative correlation
neg.cor <- subset(res.cor, abs(cor.coeff) > 0.6 & cor.coeff < 0)
head(neg.cor, 10)
#>       mirna_name target.gene_name cor.coeff    p.value       fdr status
#> 1  hsa-miR-28-5p             IER3    -0.951 0.04901988 0.1416130      V
#> 2  hsa-miR-28-5p           ZNF569    -0.939 0.06081484 0.1506698      P
#> 3  hsa-miR-28-5p         TMEM167A    -0.935 0.06511116 0.1538991      V
#> 4  hsa-miR-28-5p             E2F6    -0.920 0.07967078 0.1726200      V
#> 5  hsa-miR-28-5p             SNX1    -0.866 0.13408097 0.2582300      V
#> 6  hsa-miR-28-5p            YIPF6    -0.860 0.13971692 0.2594743      V
#> 7  hsa-miR-28-5p           TEX261    -0.834 0.16580415 0.2973040      V
#> 8  hsa-miR-28-5p          CSNK1G1    -0.827 0.17309904 0.3000383      P
#> 9  hsa-miR-28-5p           PCYOX1    -0.806 0.19422393 0.3257950      V
#> 10 hsa-miR-28-5p           HNRNPC    -0.759 0.24051212 0.3908322      V

nmir <- length(unique(neg.cor$mirna_name))
ngn <- length(unique(neg.cor$target.gene_name))
valid <- subset(neg.cor, status %in% c("PV", "V"))
```

-   Number of negative interactions: 33
-   Number of unique gene : 33
-   Number of unique miRNA : 2
-   Validated target-interaction: 12
    -   Unique genes : 12
    -   Unique miRNAs : 1


NEWS
miRTarget 1.0.1
----------------------------------------------------

NEW FEATURES

* get_mirecords_pt(): Get predicted targets for miRNAs using miRecords database. Internet connexion is required.

* get_mirecords_vt(): Download validated targets for each miRNA. Internet connexion is required.

* get_mirtarbase_vt(): Download validated targets for each miRNA using miRTarBase. Internet connexion is required.

* combine_mir_vt(): Combine miRNA validated targets from differents sources (miRecords and miRTarbase)

* combine_mir_pvt(): Combine miRNA predicted and validated targets

* combine_mir_mrna_exprs(): Combine miRNA and mRNA expression data. The output dataset will be used for statistical approach-based miRNA target predictions.

* corr_mir_mrna(): Compute correlation between miRNA and mRNA expression

* viz_mrna_per_mir(): Visualize the number of correlated mRNAs per miRNA

# To do

mir family: miFam.dat 	Family classification of miRNA hairpin sequences, http://www.mirbase.org/ftp.shtml



R/all_classes.R
#' MirMrnaExprs object and constructor
#'
#' MirMrnaExprs extends data.frame. It is used by
#' \code{\link{combine_mir_mrna_exprs}} to wrap up the results. This constructor
#' function would not typically be used by "end users".
#'
#'
#' @param data.frame a data.frame of combine_mir_mrna_exprs()
#'
#' @return a MirMrnaExprs object
#' @docType class
#' @aliases MirMrnaExprs-class
#' @rdname MirMrnaExprs
#' @export
MirMrnaExprs <- function(data.frame) {
  if(inherits(data.frame, "matrix")) data.frame <- as.data.frame(data.frame)
  structure(data.frame, class = c("data.frame", "MirMrnaExprs"))
}

#' MirTarget object and constructor
#'
#' MirTarget extends data.frame. It is used to wrap up the results
#' of predicted and validated target. This constructor
#' function would not typically be used by "end users".
#'
#'
#' @param data.frame a data frame
#'
#' @return a MirTarget object
#' @docType class
#' @aliases MirTarget-class
#' @rdname MirTarget
#' @export
MirTarget <- function(data.frame) {

   mir_cnames <- c("mirna_name", "target.gene_name",  "status")

   if(!(all(mir_cnames%in%names(data.frame))))
     stop("Target data must contains at least these three columns: ",
          paste(mir_cnames, collapse = ", "))

   structure(data.frame, class = c("data.frame", "MirTarget"))
}





R/correlation.R
#' @include all_classes.R
#' @name correlation
#' @rdname correlation
NULL
#' Correlation between miRNA and mRNA expression
#'
#' @describeIn correlation Compute correlation between miRNA and mRNA expression
#' @aliases correlation
#' @description Compute the correlation between the expression of each miRNA and
#'   its potential mRNA targets. If a gene is targeted by a miRNA we expect that
#'   the expression profile of the gene and the miRNA are anticorrelated.
#'
#' @param mir_targets an object of class MirTarget containing the miRNA-mRNA
#'   target interactions. It must contains three columns: mirna_name |
#'   target.gene_name | status.
#'
#' @param mir_mrna_exprs an object of class MirMrnaExprs, as generated by the
#'   function combine_mir_mrna_exprs(). It must contains the expression of
#'   miRNAs and mRNAs.
#' @param method method to be used for correlation analyses. Allowed values are
#'   one of "pearson", "kendall", or "spearman.
#' @param stand logical value. If TRUE the miR/mRNA expression data are
#'   standardized before computing correlation. Default value is TRUE.
#' @param mirs_of_interest a vector containing the names of miRNA for which you
#'   want to compute correlation analysis against target genes. Default value is
#'   "all".
#' @param sortby Specify how to sort the result. Allowed values are one of
#'   "cor.coeff", "p.value", "fdr", or "none": Abbreviation is allowed. Default
#'   value is "cor.coeff".
#' @param show_progress logical value. If TRUE, a progress bar is shown.
#' @return Returns an object of class CorMirTargets which is a data.frame
#'   containing correlation between miRNAs and their mRNAs. Column names of the results are as follow:
#'   \itemize{
#'   \item \bold{mirna_name}: Name of miRNAs
#'   \item \bold{target.gene_name}: The name of target genes
#'   \item \bold{cor.coeff}: Correlation coefficient
#'   \item \bold{p.value}: The p-value of the correlation test
#'   \item \bold{fdr}: Adjusted p-value using BH multiple testing correlation.
#'   This calculated separatly for the targets of each miRNA.
#'   \item \bold{status}: miRNA Target status. P for predicted, V for validated; PV for predicted and validated.
#'   }
#'
#' @author Alboukadel Kassambara \email{alboukadel.kassambara@@gmail.com}
#' @references A. Kassambara. Statistical Tools for High-throughput Data
#'   Analysis. http://www.sthda.com
#' @export
corr_mir_mrna <- function(mir_targets, mir_mrna_exprs, mirs_of_interest = "all",
                          method = "pearson", stand = TRUE,
                          sortby = c("cor.coeff", "p.value", "fdr", "none"),
                          show_progress = TRUE){

  if(!inherits(mir_mrna_exprs, "MirMrnaExprs"))
    stop("The argument mir_mrna_exprs must be an object of class MirMrnaExprs.")
  if(!inherits(mir_targets, "MirTarget"))
    stop("The argument mir_targets must be an object of class MirTarget.")

  sortby <- match.arg(sortby)[1]

  if(stand) mir_mrna_exprs <- scale(mir_mrna_exprs)

  mir_targets <- subset(mir_targets, !is.na(mir_targets$target.gene_name))
  if(mirs_of_interest =="all") mirnas <- unique(as.vector(mir_targets$mirna_name))
  else mirnas <- unique(mirs_of_interest)

  .validate_mir_names(mir_targets, mirnas)


  mir_anti_cor_target = NULL

  if(show_progress) message("\nComputing correlation between miRNAs and targets....\n")
  if(show_progress) progressBar <- txtProgressBar(min = 0, max = length(mirnas), style = 3)

  i = 0
  for(mir in mirnas){
    # miRNA target genes
    tt <- subset(mir_targets, mir_targets$mirna_name == mir)
    gns <- as.vector(tt$target.gene_name)

    # expression
    mir_exprs <- mir_mrna_exprs[, mir]
    gn_exprs <- mir_mrna_exprs[, gns, drop = FALSE ]

    # Correlation
    cor.res <- t(apply(t(gn_exprs), 1, .cortest, mir_exprs, method ))
    colnames(cor.res) <- c("cor.coeff", "p.value")
    cor.res <- as.data.frame(cor.res)
    cor.res$fdr <- p.adjust(cor.res$p.value, method = "BH")
    # cor.res <- subset(cor.res, cor.res$cor.coeff < coeff)
    if(nrow(cor.res) > 0 & sortby!="none") cor.res <- cor.res[order(cor.res[, sortby]), , drop = FALSE]
    # prepare the results
    if(nrow(cor.res)> 0) {
      rownames(cor.res) <- paste0(mir, "::", rownames(cor.res))
      mir_anti_cor_target <- rbind(mir_anti_cor_target, cor.res)
    }

    i <- i+1
    if(show_progress) setTxtProgressBar(progressBar, i)
  }

  if(!is.null(mir_anti_cor_target)){
    # mir_anti_cor_target$target.gene_name <- rownames(mir_anti_cor_target)
    mir_anti_cor_target$id <-  rownames(mir_anti_cor_target)
    mir_targets$id <- rownames(mir_targets)
    mir_anti_cor_target <- merge(mir_anti_cor_target,
                                 mir_targets,
                                 by.x = "id", by.y ="id", all.x = TRUE,
                                 sort = FALSE)
    clnames <- c("mirna_name", "target.gene_name", "cor.coeff", "p.value", "fdr", "status")
    mir_anti_cor_target <- mir_anti_cor_target[, clnames, drop = FALSE]
  }

  if(show_progress) {
    close(progressBar)
    message("**Finished**")
  }

  structure(mir_anti_cor_target, class = c("data.frame", "CorMirTarget"))
}



#' @describeIn correlation Visualize the number of correlated mRNAs per miRNA
#' @param object an object of class CorMirTarget as generated using
#' @param coeff correlation coefficient threshold. miRNAs having absolute
#' @param lab.size size of x axis labels
#' @param lab.angle rotation angle of x axis labels
#'   correlation >= coeff with one or more mRNAs are kept.
#' @export
viz_mrna_per_mir <- function(object, coeff = 0.5, lab.size = 8, lab.angle = 45,
                             output = c("plot", "table"))
  {

  output <- match.arg(output)

  if(!inherits(object, "CorMirTarget"))
    stop("An object of class CorMirTarget is required.")
  object <- subset(object, abs(object$cor.coeff) >= abs(coeff))

  # Negative and positive correlation
  neg.cor <- subset(object, object$cor.coeff < 0 )
  pos.cor <- subset(object, object$cor.coeff >  0 )

  # Count the number of positive and negative correlation per miRNA
  neg.cor <- summary(as.factor(neg.cor$mirna_name))
  neg.cor <- data.frame(mirna =  names(neg.cor), neg.cor = neg.cor)

  pos.cor <- summary(as.factor(pos.cor$mirna_name))
  pos.cor <- data.frame(mirna =  names(pos.cor), pos.cor = pos.cor)

  # Merge
  cor.count <- merge(neg.cor, pos.cor, by.x = "mirna", by.y = "mirna",
                     all = TRUE, sort = FALSE)
  cor.count$neg.cor[is.na(cor.count$neg.cor)]  <- 0
  cor.count$pos.cor[is.na(cor.count$pos.cor)]  <- 0
  cor.count$tot <- cor.count$neg.cor + cor.count$pos.cor
  cor.count <- cor.count[order(cor.count$tot, decreasing = TRUE), ]

  if(output == "table") return(cor.count)
  # Melt
  cor.count2 <- reshape2::melt(cor.count, measure.vars = c("neg.cor", "pos.cor"))
  cor.count2$value[is.na(cor.count2$value)] <- 0
  cor.count2$variable <- relevel(cor.count2$variable, "neg.cor")
  # cor.count2 <- cor.count2[order(cor.count2$value, decreasing = TRUE), ]
  ggplot2::ggplot(cor.count2, ggplot2::aes_string("mirna", "value")) +
    ggplot2::geom_bar(stat = "identity", ggplot2::aes_string(fill = "variable"))+
    ggplot2::theme_minimal() +
    ggplot2::scale_x_discrete(limits = cor.count$mirna)+
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle=lab.angle, size = lab.size)) +
    ggplot2::labs(y = "Number of coexpressed mRNAs", x = NULL, fill = "Count")
}


#' @describeIn correlation Fraction of negative correlation events
#' @param output allowed values are "plot" or "table"
#' @export
viz_neg_cor_fraction <- function(object, output = c("plot", "table")){
  coeff <- seq(from = 0, to = 0.95, by = 0.05)
  output <- match.arg(output)

  neg.frac <- NULL
  for(.coeff in coeff){
    res.count <- viz_mrna_per_mir(object, coeff = .coeff, output = "table")
    neg.frac <- c(neg.frac, sum(res.count$neg.cor)/sum(res.count$tot))
  }

  res <- data.frame(coeff = coeff, neg.frac = neg.frac)
  if(output == "table") return(res)
  plot(res, xlab = "Threshold of absolute correlation",
       ylab = "Fraction of negative co-expression events.",
       type = "b")

}




# A helper function
.cortest <- function(x, y, method = "pearson"){
  x <- as.numeric(x)
  y <- as.numeric(y)
  res <- cor.test(x, y, method = method)
  c(round(res$estimate,3), res$p.value)
}

# Check if miRNA names provided by users is valid
.validate_mir_names <- function(mir_targets, names){
  dd <- setdiff(names, mir_targets$mirna_name)
  if(length(dd)!=0) stop("Can't found the following miRNA names in your data: ",
                         paste(dd, collapse =", "))
}



R/combine_mir_vt.R
#' @include all_classes.R
NULL
#'Combine miRNA validated targets from differents sources (miRecords and miRTarbase)
#'
#'@description A helper function for combining miRNA targets from differents
#'  sources. Duplicated miRNA-target interactions are automatically removed.
#'
#'@param ... a list of MirTarget object to combine. MirTarget object can be
#'  generated by the functions get_mirecords_*() or by get_mirtarbase_*().
#' @return  Returns an
#'   object of class MirTarget which is a data frame containing combined validated
#'   targets for each miRNA.
#'
#'@author Alboukadel Kassambara \email{alboukadel.kassambara@@gmail.com}
#'@references A. Kassambara. Statistical Tools for High-throughput Data
#'  Analysis. http://www.sthda.com
#' @examples
#' \dontrun{
#' # Example of miRNAs
#' miRNAs <- c("hsa-miR-16", "hsa-miR-155")
#'
#' # Validated target from miRecords
#' mirecords_vt <- get_mirecords_vt(miRNAs)
#'
#'# Validated target from miRTarbase
#' mirtarbase_vt <- get_mirtarbase_vt(miRNAs)
#'
#' v_targets <- combine_mir_vt(mirecords_vt, mirtarbase_vt)
#' head(v_targets)
#' }
#'
#'@name combine_mir_vt
#'@rdname combine_mir_vt
#'@export
combine_mir_vt <- function(...){

  objects <- list(...)

  fargs <- as.list(match.call(expand.dots = TRUE))
  objnames <- as.character(fargs)[-1]

  for (i in 1:length(objects)) {
    if(!inherits(objects[[i]], "MirTarget"))
      stop(objnames[[i]], "must be an object of class MirTarget.")
  }
  res <- objects[[1]]

  for (i in 2:length(objects)) {
    res <- rbind(res, objects[[i]])
  }
  # Remove duplicated rows
  row_ids <- paste0(res$mirna_name, res$target.gene_name)
  res <- res[!duplicated(row_ids), ]
  res <- na.omit(res)
  MirTarget(res)
}


R/combine_mir_pvt.R
#' @include all_classes.R
NULL
#'Combine miRNA predicted and validated targets
#'
#'@description A helper function for combining miRNA predicted and validated
#'  targets. Duplicated miRNA-target interactions are automatically removed.
#'@param p_targets,v_targets an object of class MirTarget
#'
#'@return  Returns an object of class MirTarget which is a data frame containing
#'  combined predicted and validated targets for each miRNA. The final results
#'  might contains different status of miRNA-target interactions (MTIs):
#'  \itemize{ \item V: for validated MTIs \item P: for predicted MTIs}
#'
#'@author Alboukadel Kassambara \email{alboukadel.kassambara@@gmail.com}
#'@references A. Kassambara. Statistical Tools for High-throughput Data
#'  Analysis. http://www.sthda.com
#' @examples
#' \dontrun{
#' # Example of miRNAs
#' miRNAs <- c("hsa-miR-16", "hsa-miR-155")
#'
#' # Validated target from miRecords
#' v_targets <- get_mirecords_vt(miRNAs)
#'
#' # Get predicted target
#' p_targets <- get_mirecords_pt(miRNAs)
#'
#'# Combine
#' targets <- combine_mir_pvt(p_targets, v_targets)
#' head(targets)
#' }
#'
#'@name combine_mir_pvt
#'@rdname combine_mir_pvt
#'@export
combine_mir_pvt <- function(p_targets, v_targets){

  if(!inherits(p_targets, "MirTarget"))
    stop("p_targets must be an object of class MirTarget.")

  if(!inherits(v_targets, "MirTarget"))
    stop("v_targets must be an object of class MirTarget.")

  p_targets$status <- rep("P", nrow(p_targets))
  v_targets$status <- rep("V", nrow(v_targets))

  rownames(v_targets) <- paste0(v_targets$mirna_name, "::", v_targets$target.gene_name)
  rownames(p_targets) <- paste0(p_targets$mirna_name, "::", p_targets$target.gene_name)

  only_v <- subset(v_targets, !(rownames(v_targets) %in% rownames(p_targets)))
  only_p <- subset(p_targets, !(rownames(p_targets) %in% rownames(v_targets)))
  pv <- subset(p_targets, (rownames(p_targets) %in% rownames(v_targets)))
  pv$status <- rep("V", nrow(pv))
  targets <- rbind(only_v, only_p, pv)
  rownames(targets) <- 1:nrow(targets)
  targets <- na.omit(targets)
  MirTarget(targets)
}


R/get_mirecords.R
#' @include all_classes.R
NULL
#' Get predicted and validated targets for miRNAs using miRecords database
#'
#' @description miRecords (\url{http://c1.accurascience.com/miRecords/}) is a
#'   resource for animal miRNA-target interactions. There are two components in
#'   miRecords. The experimentally Validated Targets (vt) component and the
#'   Predicted Targets (pt) component which is an integration of predicted miRNA
#'   targets produced by 11 established miRNA target prediction programs.
#'   \itemize{
#'
#'   \item \bold{get_mirecords_pt()}: Recursively retrieve the predicted targets
#'   for each miRNA. Only predicted targets by at least n/11 database are kept,
#'   where \bold{n} is user-defined. Internet connexion is required the first
#'   time you use this function. \item \bold{get_mirecords_vt()}: Download
#'   validated targets for each miRNA. Internet connexion is required. }
#'
#' @param miRNAs a character vector containing miRNA names (e.g.: miRNAs =
#'   c("hsa-miR-16", "hsa-miR-34a))
#' @param n integer specifying the minimal number of prediction for each miRNA.
#'   In other words, Only predicted target by at least n/11 database are kept.
#'   Default value is 5.
#' @param species species names as available on miRecords database. Possible
#'   values are one of: "Homo sapiens", "Mus musculus", "Rattus norvegicus",
#'   "Caenorhabditis elegans", "Drosophila melanogaster", "Gallus gallus", "Ovis
#'   aries", "Canis familiaris". Default value is "Homo sapiens".
#' @param result_file path to result file for holding miRNA predicted/validated
#'   targets. Default values are "mirna_mirecords_predicted_targets.txt".
#' @param keep_db_names logical value. If TRUE, keep the name of the 11 databases used for target predictions.
#' @param show_progress logical value. If TRUE, a progress bar is shown.
#' @return \itemize{ \item \bold{get_mirecords_pt()}: Downloads and returns an object of class MirTarget which is a
#' data frame containing predicted targets for each miRNA. It saves also the
#' results as a tab-delimited .txt file, which default name is
#' "mirna_mirecords_predicted_targets.txt". If this file exists already in your
#' working directory, then get_mirecords_pt() will simply import it into R.
#' \item \bold{get_mirecords_vt()}:  Downloads and returns an object of class MirTarget which is a data frame
#' containing validated targets for each miRNA.
#'
#' }
#' @author Alboukadel Kassambara \email{alboukadel.kassambara@@gmail.com}
#' @references A. Kassambara. Statistical Tools for High-throughput Data
#'   Analysis. http://www.sthda.com
#' @examples
#' \dontrun{
#' # Predicted targets
#'  ptget <- get_mirecords_pt(c("hsa-miR-16", "hsa-miR-155"))
#'  head(ptget)
#'
#' # Validated targets
#'  vtget <- get_mirecords_vt(c("hsa-miR-16", "hsa-miR-155"))
#'  head(vtget)
#'
#' }
#'
#' @name get_mirecords
#' @rdname get_mirecords
#' @export
get_mirecords_pt <- function(miRNAs, n = 5, species = "Homo sapiens",
                             result_file = "mirna_mirecords_predicted_targets.txt",
                             keep_db_names = FALSE,
                             show_progress = TRUE){

  .check_species(species)

  # Columns contained in the result file

  columns <- c("miRNA ID", 	"Symbol",
               "diana", "microinspector",	"miranda",	"mirtarget2",
               "mitarget", "nbmirtar",	"pictar", "pita",	"rna22",
               "rnahybrid", "targetscan")

  columns_2 <- columns
  columns_2[1:2] <- c("mirna_name", "target.gene_name")
  species = gsub(" ", "+", species, fixed = TRUE)

  if(file.exists(result_file)){
    warning(result_file, " exists already and has been simply imported into R. ",
             "If you want to download it again, remove the old file.")

    res <- read.table(result_file, sep ="\t", header= TRUE, row.names = NULL,
                      comment.char = "#", quote = "")
    rsum <- apply(res[, 3:13], 1, sum)
    res <- subset(res, rsum >= n)

    # Remove duplicated rows
    row_ids <- paste0(res$mirna_name, res$target.gene_name)
    res <- res[!duplicated(row_ids), , drop = FALSE]

    if(!keep_db_names) res <- res[, 1:2, drop = FALSE]

    res$status <- rep("predicted", nrow(res))
    res <- subset(res, res$mirna_name %in% miRNAs)
    return(MirTarget(res))
  }

    write.table(matrix(columns_2, nrow = 1), file = result_file, sep ="\t",
              append = FALSE, quote = FALSE, row.names = FALSE,
              col.names = FALSE)

  # For each miRNA find the predicted target by at least 5 data base
  if(show_progress) message("\nDownloading predicted target....\n")
  if(show_progress) progressBar <- txtProgressBar(min = 0, max = length(miRNAs), style = 3)

  i = 0
  for(mir in miRNAs){
    # 1. Download predicted target for the miRNA
    url <- paste0("http://c1.accurascience.com/miRecords/download_data.php?",
                  "p=1&species=", species, "&targetgene_type=refseq_acc&",
                  "targetgene_info=&search_int=Search&pn=", n, "&mirna_acc=", mir)

    dest_file <- paste0(mir, ".xls")
    download.file(url, dest_file, quiet = TRUE)

    # 2. Read the xls file
    dd <- read.table(dest_file, sep="\t",
                     header = TRUE, row.names = NULL, quote = "",
                     comment.char = "#", check.names = FALSE)
    dd <- dd[, columns, drop = FALSE]
    if(nrow(dd)== 0) dd[1, ] <-c(mir, rep(NA, length(columns)-1))

    # 3. Append dd in the final output file
    write.table(dd, file = result_file, sep ="\t",
                append = TRUE, quote = FALSE, row.names = FALSE,
                col.names = FALSE)

    # unlink xlsx file
     unlink(dest_file)
    i = i+1
    if(show_progress) setTxtProgressBar(progressBar, i)
  }
  if(show_progress) {
    close(progressBar)
    message("**Finished**")
  }
  message("Results have been saved in the following file: ", result_file)
  res <- read.table(result_file, sep ="\t", header= TRUE, row.names = NULL,
                    quote = "", comment.char = "#")
  # Remove duplicated rows
  row_ids <- paste0(res$mirna_name, res$target.gene_name)
  res <- res[!duplicated(row_ids), , drop = FALSE]

  if(!keep_db_names) res <- res[, 1:2, drop = FALSE]

  res$status <- rep("predicted", nrow(res))
  return(MirTarget(res))
}

#' @param version miRecords version. Default value is 4
#' @name get_mirecords
#' @rdname get_mirecords
#' @export
get_mirecords_vt <- function(miRNAs, species = "Homo sapiens", version = 4){

  .check_species(species)

  message("Downloading miRecords miRNA validated targets...")

  url <- paste0("http://c1.accurascience.com/miRecords/download_data.php?v=", version)
  download.file(url, "mirecords_vt.xls", quiet = TRUE)
  mirecords <- readxl::read_excel("mirecords_vt.xls")
  unlink("mirecords_vt.xls")
  hs_vt <- mirecords[which(mirecords$miRNA_species %in% species),
                     c("Target gene_name",  "miRNA_mature_ID")]
  colnames(hs_vt) <- tolower(make.names(colnames(hs_vt)))

  miRNAs <- data.frame(mirna_name = miRNAs)
  v_target <- merge(x=miRNAs, y = hs_vt,
                    by.x="mirna_name", by.y="mirna_mature_id",
                    all.x = TRUE, sort = FALSE)
  # Remove duplicated rows
  row_ids <- paste0(v_target$mirna_name, v_target$target.gene_name)
  v_target <- v_target[!duplicated(row_ids), ]

  v_target$status <- rep("validated", nrow(v_target))
  return(MirTarget(v_target))
}


# Helper functions
.check_species <- function(species)
  {
  sp <- c("Homo sapiens", "Mus musculus", "Rattus norvegicus",
          "Caenorhabditis elegans", "Drosophila melanogaster",
          "Gallus gallus", "Ovis aries", "Canis familiaris")
  if(!(species %in% sp)) stop(species, " is not supported. Allowed values are: ",
                              paste(sp, collapse =", "))
 }
